smartsim/_core/launcher/dragon/pqueue.py

```diff
@@ -0,0 +1,344 @@
+
+    def unassigned(
+        self, heap: t.Optional[t.List[_NodeRefCount]] = None
+    ) -> t.List[_NodeRefCount]:
+        """Select nodes that are currently not assigned a task
+
+        :param heap: (optional) a subset of the node heap to consider
+        :returns: a list of reference counts for all unassigned nodes"""
+        if heap is None:
+            return [node for node in self._ref_map.values() if node[0] == 0]
+
+        return [node for node in heap if node[0] == 0]
+
+    def assigned(
+        self, heap: t.Optional[t.List[_NodeRefCount]] = None
+    ) -> t.List[_NodeRefCount]:
+        """Helper method to identify the nodes that are currently assigned
+
+        :param heap: (optional) a subset of the node heap to consider
+        :returns: a list of reference counts for all assigned nodes"""
+        if heap is None:
+            return [node for node in self._ref_map.values() if node[0] == 1]
+
+        return [node for node in heap if node[0] == 1]
+
+    def _check_satisfiable_n(
+        self, num_items: int, heap: t.Optional[t.List[_NodeRefCount]] = None
+    ) -> bool:
+        """Validates that a request for some number of nodes `n` can be
+        satisfied by the prioritizer given the set of nodes available
+
+        :param num_items: the desird number of nodes to allocate
+        :param heap: (optional) a subset of the node heap to consider"""
+        num_nodes = len(self._ref_map.keys())
+
+        if num_items < 1:
+            msg = "Cannot handle request; request requires a positive integer"
+            logger.warning(msg)
+            return False
+
+        if num_nodes < num_items:
+            msg = f"Cannot satisfy request for {num_items} nodes; {num_nodes} in pool"
+            logger.warning(msg)
+            return False
+
+        num_open = len(self.unassigned(heap))
+        if num_open < num_items:
+            msg = f"Cannot satisfy request for {num_items} nodes; {num_open} available"
+            logger.warning(msg)
+            return False
+
+        return True
+
+    def _get_next_available_node(
+        self, heap: t.List[_NodeRefCount]
+    ) -> t.Optional[_NodeRefCount]:
+        """Finds the next node w/the least amount of running processes and
+        ensures that any elements that were directly updated are updated in
+        the priority structure before being made available
+
+        :param heap: (optional) a subset of the node heap to consider
+        :returns: a reference counter for an available node if an unassigned node
+        exists, `None` otherwise"""
+        tracking_info: t.Optional[_NodeRefCount] = None
+
+        with self._lock:
+            tracking_info = heapq.heappop(heap)
+            is_dirty = tracking_info[2]
+
+            while is_dirty:
+                if is_dirty:
+                    # mark dirty items clean and place back into heap to be sorted
+                    tracking_info[2] = 0
+                    heapq.heappush(heap, tracking_info)
```

Comment:
Great suggestion. The while loop was annoying. I rewrote the method to mark all nodes as clean then rebuild the heap a single time instead of using a loop to step through the dirty items. It's more reliable and simpler to read, now.